probe-rs/src/bin/probe-rs/cmd/run/test_run_mode.rs

```diff
@@ -0,0 +1,292 @@
+    pub format: Option<FormatSetting>,
+
+    /// If set, filters are matched exactly rather than by substring.
+    #[clap(long = "exact", help_heading = "TEST OPTIONS")]
+    pub exact: bool,
+
+    /// Options which are ignored, but exist for compatibility with libtest.
+    /// E.g. so that vscode and intellij can invoke the test runner with the args they are used to
+    #[clap(flatten)]
+    _no_op: NoOpTestOptions,
+}
+
+/// Options which are ignored, but exist for compatibility with libtest.
+#[derive(Debug, clap::Parser)]
+struct NoOpTestOptions {
+    // No-op, ignored (libtest-mimic always runs in no-capture mode)
+    #[clap(long = "nocapture", hide = true)]
+    nocapture: bool,
+
+    /// No-op, ignored. libtest-mimic does not currently capture stdout.
+    #[clap(long = "show-output", hide = true)]
+    show_output: bool,
+
+    /// No-op, ignored. Flag only exists for CLI compatibility with libtest.
+    #[clap(short = 'Z', hide = true)]
+    unstable_flags: Option<String>,
+}
+
+/// Test run mode
+pub struct TestRunMode {
+    libtest_args: Arguments,
+}
+
+impl TestRunMode {
+    pub fn new(test_options: &TestOptions) -> Box<Self> {
+        Box::new(Self {
+            libtest_args: Arguments {
+                test_threads: Some(1), // Avoid parallel execution
+                list: test_options.list,
+                exact: test_options.exact,
+                format: test_options.format,
+                filter: if test_options.filter.is_empty() {
+                    None
+                } else {
+                    //TODO: Fix libtest-mimic so that it allows multiple filters (same as std test runners)
+                    Some(test_options.filter.join(" "))
+                },
+                ..Arguments::default()
+            },
+        })
+    }
+
+    /// Asks the target for the tests, and create a "run the test"-closure for each test.
+    /// libtest-mimic is in charge of selecting the tests to run based on the filter and other options
+    fn create_tests(session_and_runloop_ref: Arc<Mutex<SessionAndRunLoop>>) -> Result<Vec<Trial>> {
+        let mut session_and_runloop = session_and_runloop_ref.lock().unwrap();
+        let list = Self::list_tests(&mut session_and_runloop)?;
+
+        let mut tests = Vec::<Trial>::new();
+        for t in &list.tests {
+            let test = t.clone();
+            let session_and_runloop = session_and_runloop_ref.clone();
+            tests.push(
+                Trial::test(&t.name, move || {
+                    let mut session_and_runloop = session_and_runloop.lock().unwrap();
+                    Self::run_test(test, &mut session_and_runloop)
+                })
+                .with_ignored_flag(t.ignored),
+            )
+        }
+        Ok(tests)
+    }
+
+    const SEMIHOSTING_USER_LIST: u32 = 0x100;
+
+    /// Requests all tests from the target via Semihosting back and forth
+    fn list_tests(session_and_runloop: &mut SessionAndRunLoop) -> Result<Tests> {
```

Comment:
Not sure if I understand this correctly, but it looks like testing needs the ELF to be specified. Why are the test cases listed in runtime and not encoded into some metadata in the ELF?

Reply:
Yes, this is an optimization I have in mind for embedded-test 2.0. Currently the test metadata is exchanged between target and probe-rs via semihosting / JSON :disappointed: .

I've created embedded-test because defmt-test did not support riscv or async test functions. Back then, I just jumped on the semihosting train and did not think about storing the metadata in the ELF.  But now, that we have an `embedded-test.x` linker file anyway, this is the obvious way to go. 

Unfortunately, my current client has decided to limit our testing effort to the bare minimum, as a consequence my time to work on embedded-test is currently limited to my leisure time and this refactoring will have to wait until I'm less busy.